File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

public class Solution {

    // 평균 반환 시간을 최소화하는 우선순위 디스크 컨트롤러 메서드
    public static int solution(int[][] jobs) {
        // 현재 시간
        int now = 0;

        // 처리된 작업 수
        int count = 0;

        // 마지막으로 작업이 끝난 시점
        int start = -1;

        // 총 반환 시간
        int totalTime = 0;

        // 작업 대기 큐: 작업 소요시간 기준 오름차순 정렬
        PriorityQueue<int[]> heap = new PriorityQueue<>(
            (a, b) -> Integer.compare(a[1], b[1])
        );

        // jobs 배열을 리스트로 변환하여 처리
        List<int[]> jobList = Arrays.asList(jobs);
        
        // 전체 작업이 처리될 때까지 반복
        while (count < jobs.length) {
            // 현재 시점까지 요청된 작업들을 힙에 추가
            for (int[] job : jobs) {
                int request = job[0];
                int duration = job[1];

                if (start < request && request <= now) {
                    heap.offer(new int[]{request, duration});
                }
            }

            // 처리할 작업이 있다면
            if (!heap.isEmpty()) {
                // 가장 짧은 작업 소요시간을 가진 작업 꺼냄
                int[] current = heap.poll();

                // 요청 시각
                int requestTime = current[0];

                // 소요 시간
                int durationTime = current[1];

                // 작업 시작 시점 갱신
                start = now;

                // 현재 시간 증가 (작업 실행)
                now += durationTime;

                // 반환 시간 누적 (종료 시점 - 요청 시점)
                totalTime += now - requestTime;

                // 처리된 작업 수 증가
                count++;
            } else {
                // 대기 작업이 없으면 시간 증가 (idle)
                now++;
            }
        }

        // 평균 반환 시간의 정수 부분 반환
        return totalTime / jobs.length;
    }

    // 테스트 케이스 실행을 위한 main 메서드
    public static void main(String[] args) {
        int[][] jobs = {{0, 3}, {1, 9}, {3, 5}};
        System.out.println(solution(jobs)); // 출력: 8
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
// Heap 클래스 정의 (Min-Heap)
class Heap {
    // 힙 데이터 저장 배열
    #heap = [];

    // 값 추가
    push(value) {
        this.#heap.push(value);
        this.#bubbleUp();
    }

    // 최상위 요소 제거 및 반환
    pop() {
        if (this.#heap.length === 0) return null;
        if (this.#heap.length === 1) return this.#heap.pop();

        const top = this.#heap[0];
        this.#heap[0] = this.#heap.pop();
        this.#sinkDown();
        return top;
    }

    // 최상위 요소 반환 (삭제 안 함)
    peek() {
        return this.#heap[0] ?? null;
    }

    // 현재 힙 크기 반환
    size() {
        return this.#heap.length;
    }

    // 내부 메서드: 위로 올리기
    #bubbleUp() {
        let index = this.#heap.length - 1;
        const value = this.#heap[index];

        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            const parent = this.#heap[parentIndex];

            if (value[0] >= parent[0]) break;

            this.#heap[index] = parent;
            this.#heap[parentIndex] = value;
            index = parentIndex;
        }
    }

    // 내부 메서드: 아래로 내리기
    #sinkDown() {
        let index = 0;
        const length = this.#heap.length;
        const value = this.#heap[0];

        while (true) {
            const leftIndex = 2 * index + 1;
            const rightIndex = 2 * index + 2;
            let smallest = index;

            if (leftIndex < length && this.#heap[leftIndex][0] < this.#heap[smallest][0]) {
                smallest = leftIndex;
            }

            if (rightIndex < length && this.#heap[rightIndex][0] < this.#heap[smallest][0]) {
                smallest = rightIndex;
            }

            if (smallest === index) break;

            this.#heap[index] = this.#heap[smallest];
            this.#heap[smallest] = value;
            index = smallest;
        }
    }
}

// 디스크 컨트롤러 문제 해결 함수
function solution(jobs) {
    // 총 반환 시간 누적 변수
    let answer = 0;

    // 현재 시간
    let now = 0;

    // 처리된 작업 수
    let i = 0;

    // 마지막 작업 시작 시각
    let start = -1;

    // 최소 힙 생성
    const heap = new Heap();

    // 모든 작업을 처리할 때까지 반복
    while (i < jobs.length) {
        // 현재 시간까지 요청된 작업을 힙에 추가
        for (let job of jobs) {
            const [request, duration] = job;

            if (start < request && request <= now) {
                // (소요 시간, 요청 시간) 형식으로 저장
                heap.push([duration, request]);
            }
        }

        // 힙에 처리할 작업이 있는 경우
        if (heap.size() > 0) {
            // 소요 시간이 가장 짧은 작업 꺼내기
            const [duration, request] = heap.pop();

            // 작업 시작 시간 업데이트
            start = now;

            // 현재 시간 증가
            now += duration;

            // 반환 시간 누적
            answer += now - request;

            // 처리된 작업 수 증가
            i += 1;

        } else {
            // 대기 작업이 없으면 현재 시간 1 증가
            now += 1;
        }
    }

    // 평균 반환 시간의 정수 부분 반환
    return Math.floor(answer / jobs.length);
}

// 테스트 실행
console.log(solution([[0, 3], [1, 9], [3, 5]])); // 8